code/modules/mob/living/living_movement.dm

```diff
@@ -1,7 +1,58 @@
 /mob/living/Moved(atom/old_loc, movement_dir, forced, list/old_locs, momentum_change = TRUE)
 	. = ..()
 	update_turf_movespeed(loc)
+	if(HAS_TRAIT(src, TRAIT_NEGATES_GRAVITY))
+		if(!isgroundlessturf(loc))
+			ADD_TRAIT(src, TRAIT_IGNORING_GRAVITY, IGNORING_GRAVITY_NEGATION)
+		else
+			REMOVE_TRAIT(src, TRAIT_IGNORING_GRAVITY, IGNORING_GRAVITY_NEGATION)
+
+	var/turf/old_turf = get_turf(old_loc)
+	var/turf/new_turf = get_turf(src)
+	// If we're moving to/from nullspace, refresh
+	// Easier then adding nullchecks to all this shit, and technically right since a null turf means nograv
+	if(!old_turf || !new_turf)
```

Comment:
```suggestion
	if(isnull(old_turf) || isnull(new_turf))
```

couldn't we do this for sanic speeds?

Reply:
the isnull lobby wants to ruin my life

Reply:
fyi the main win of this is explicitness [and sometimes correctness], the sanic speed is very marginal but nice